R/ChromBackendSpectra.R

```diff
@@ -0,0 +1,254 @@
+                  # automatically added. and for mzmin mzmax evenif the columns
+                  # are automaticall created,
+                  # the code later WILL FAIL if the input is NA.
+                  # Not sure how to handle that..
+              }
+
+              f <- factor(
+                  do.call(
+                      paste,
+                      c(as.list(Spectra::spectraData(spectra)[, factorize.by]),
+                        sep = "_")))
+              spectra$chromSpectraIndex <- f
+              full_sp <- do.call(rbindFill,
+                                 lapply(split(spectra, f),
+                                        .spectra_format_chromData))
+              full_sp$chromIndex <- seq_len(nrow(full_sp))
+              full_sp$chromSpectraIndex <- levels(f)
+              rownames(full_sp) <- NULL
+              object@spectra <- spectra
+              callNextMethod(object, chromData = full_sp)
+              }
+          )
+
+#' @rdname hidden_aliases
+#' @importFrom methods callNextMethod
+setMethod("show", "ChromBackendSpectra", function(object) {
+    callNextMethod()
+    cat("\nThe Spectra object contains", length(object@spectra), "spectra\n")
+    if (object@inMemory) cat("\nPeaks data is cached in memory\n")
+})
+
+#' @rdname ChromBackendSpectra
+#' @note
+#' Does it needs to be a generic and implemented for ChromBackend ? It's very
+#' specific for this backend, also I would NOT allow replacement of it.
+#' @export
+chromSpectraIndex <- function(object) {
+    if (!is(object, "ChromBackendSpectra"))
+        stop("The object must be a 'ChromBackendSpectra' object.")
+    chromData(object, columns = "chromSpectraIndex", drop = TRUE)
+}
+
+#' @rdname hidden_aliases
+setMethod("factorize", "ChromBackendSpectra",
+          function(object, factorize.by = c("msLevel", "dataOrigin"),...) {
+              if (!all(factorize.by %in% chromVariables(object)))
+                  stop("All 'factorize.by' variables must be in chromData.")
+              if (!all(factorize.by %in% Spectra::spectraVariables(object@spectra)))
+                  stop("All 'factorize.by' variables must be in the ",
+                       "Spectra object.")
+              object@chromData$chromSpectraIndex <- factor(do.call(
+                  paste, c(object@chromData[, factorize.by], sep = "_")))
+              object@spectra$chromSpectraIndex <- factor(
+                  do.call(
+                      paste,
+                      c(as.list(Spectra::spectraData(object@spectra)[, factorize.by]),
+                        sep = "_")),
+                  levels = levels(chromSpectraIndex(object)))
+              object
+          })
+
+#' @rdname hidden_aliases
+#' @importMethodsFrom ProtGenerics backendParallelFactor
+setMethod("backendParallelFactor", "ChromBackendSpectra", function(object, ...) {
+    factor(chromSpectraIndex(object), levels = unique(chromSpectraIndex(object))) # now sure about that.
```

Comment:
hm. also not sure. do you need to implement a method for that now? I would maybe just return `factor()`, so for now no parallel processing by default.